### PR TITLE
Implement extended serveinfo protocol Fix #2135

### LINKED ISSUE #2135: 64 players: ServerInfo might be too big
The ServerInfo packet size can be greater than the the max allowed packet size.
Basically, a ServerInfo packet is composed by :
```
 - header                 (18)
   - packet flags (1)
   - client token (4)
   - server token (4)
   - request data (8)
   - null byte    (1)
 - server properties (max 410)
   - version       (max 32)
   - name          (max 64)
   - hostname     (max 256)
   - map           (max 32)
   - gametype      (max 16)
   - password flag      (1)
   - skill level        (1)
   - num_player     (max 2)
   - max_player     (max 2)
   - num_client     (max 2)
   - max_client     (max 2)
 - player info        (max 36) x64
   - name    (max 16)
   - clan    (max 12)
   - country  (max 2)
   - score    (max 5)
   - flag         (1)

Total: max 2732 >> 1400
```

**Step to reproduce:**
Fill a 64 clients server with player names and clan names of max size.
` [netserver]: packet payload too big. 2035. dropping packet`

### PR DESCRIPTION
**Pull request by Learath2**
There is an off by one error here, I'm sending 32 clients the client thinks it receives 33.

I'm having trouble debugging it and it's getting on my nerves so I'm leaving this here.

### CODE REVIEW

**Comment by Learath2** [comment]
@heinrich5991 you are the one that came up with the protocol, maybe you could figure out what I'm doing wrong.

**Comment by jxsl13** [comment]
What's the purpose of this pr?

**Comment by Dune-jr** [comment]
To fix https://github.com/teeworlds/teeworlds/issues/2135

**Comment by Learath2** [comment]
Okay, works now.

It should be tested by a couple people before merging though.
@fokkonaut you run a server that's frequently full you can try.

**Comment by nheir** [comment]
I got 23/23 with a full 64/64 server.

screenshot_2020-03-09_23-00-31

btw, is it be possible to stay compatible with 0.7.3 clients ?

nb: there is some debug code in the PR (`true || ...`, `//dbg_assert...`)

**Comment by Learath2** [comment]
Oh forgot those. The `dbg_assert` was to fix `dbg_dummies`. The if statement was to fake the serverinfo.

23/23 doesn't really make sense to me, help me reproduce it and I'll fix it.

**Comment by nheir** [comment]
I tried with server bots, but I might add the bug while merging. 
Anyway, you can give a try with this branch:
https://github.com/nheir/teeworlds/tree/tmp
(build with `bam`)
```bash
$ build/x86_64/release/teeworlds_srv 'sv_max_clients 64; sv_bot_slots 64; sv_player_slots 64; sv_register 0'
```

**Comment by Learath2** [comment]
@nheir it's 64/64 for me. Can you reproduce this consistently?

@heinrich5991 what is the best way to keep this backwards compatible? Is there any indication in the request that the client is old? Should I introduce a newer request packet?

@oy do we need to keep backwards compatibility?

File: src/engine/client/client.cpp [hunk 1]
```diff
@@ -954,9 +992,11 @@ int CClient::UnpackServerInfo(CUnpacker *pUnpacker, CServerInfo *pInfo, int *pTo
 			NumPlayers++;
 	}
 	pInfo->m_NumPlayers = NumPlayers;
```
Comment (nheir) [inline src/engine/client/client.cpp]: `NumPlayers` is misused. Maybe remove this line ?

Comment (Learath2) [reply src/engine/client/client.cpp]: Ah, yep I missed that one

**Comment by Learath2** [comment]
I don't like the way this turned out, I'll give it another go with backwards compatibility in mind

**Pull request closed without merging**